### PR TITLE
Fixes Issue#325 - UI and Documentation Improvements

### PR DESCRIPTION
**Pull request by TabLand**
### UI Improvements for Selection Tools
    
Make it easier to exit selection tools when presented with an empty advance
selection. Also add support for warning messages.

### Documentation improvements
    

1) 4 point move tool: point G should be point C
2) Lengthen crease line (Specified color) tool: Help is hidden due to a syntax typo

### CODE REVIEW

**Comment by TabLand** [comment]
#325

**Comment by Minh-Than** [comment]
I guess you already worked on this issue ahead of me haha. I was trying to fix it by just straight up disabling the move/copy tools when there's no prior selection being made. But maybe what you're doing is a good idea.

**Comment by TabLand** [comment]
Hi @Minh-Than,

There's certainly merits in the option you've suggested. 

For the disable / enable button UI experience to be fluid - I think it makes sense to start caching the selected line segments, and fire an event on selection change? Something like this could be reused by other tools. The warning wording could also be improved e.g "Selection Transformation tools require crease(s) to be selected in advance"; and then the help text doesn't need to be displayed until the buttons are activated.

Would appreciate your thoughts,

**Comment by Minh-Than** [comment]
Admittedly, I'm still very new to Java and especially the architecture of Edita, so what I did to implement the enabling and disabling them is quite crude at the very least and it's certainly not considering what you said like caching line segments for other tools to use. But I did get it to work.

At the moment I got these:

https://github.com/oriedita/oriedita/assets/94136126/0d0605d9-8a2f-4a32-ba7f-9e083265dc04


https://github.com/oriedita/oriedita/assets/94136126/c8b7b128-ba54-4a27-9c6f-a388e7667fb2

**Comment by TabLand** [comment]
That does look pretty neat. 

If you could push these changes to a branch in your repository fork, I can try to merge them into this pull request. I'm also interested in testing your changes against a larger crease pattern; the performance improvements I had in mind might not be immediately needed.

**Comment by Minh-Than** [comment]
I tested on Ryujin 3.5 with neck twist CP and it seems to be fine, keeping about the same performance, but could be biased. I was actually planning to wait for your PR to be merged since the only conflict is in left panel and i only wrote a new setData() and one line of code. But your proposal seems to make more sense so I will do that.

Should be up [here](https://github.com/Minh-Than/oriedita/tree/disable_move_copy_when_no_crease_select).

**Comment by TabLand** [comment]
The event driven code is very similar to the concept I had. Will have a look into merging this later tonight

The only thing that stands out is the use of getFoldLineTotalForSelectFolding() method in FoldLineSet - I was reviewing this yesterday and this function excludes counting auxillary lines given they are not relevant to folding. However, these are in scope for the selection transformation tools, hence a new counting function will be needed in CreasePattern_Worker + FoldLineSet.

Thanks for sharing your commit, it saves a lot of time 🙂️

**Comment by Minh-Than** [comment]
Oh I completely forgot that that method ignores the aux lines haha. Well I'm glad it was something similar to what you have in mind at least. Wish you best in figuring out the rest to fix this issue.

**Comment by Minh-Than** [comment]
I realized that writing a new method for that was pretty easy so quickly made a [commit](https://github.com/oriedita/oriedita/commit/6b74945137fa428e8382dfd87946f74f152eeba3) for you to look at in case you haven't done it already.

**Comment by TabLand** [comment]
Very close to pushing my changes, just tied up with further testing and cleanup.
Your change is fairly close to mine, I chose to return true if a single line is found to be [selected](https://github.com/oriedita/oriedita/commit/6b74945137fa428e8382dfd87946f74f152eeba3#diff-73ccc3efe4a5a8ceb1308fdbcf35773eb47fcc813167228df6d0d0cc4d56cd6fR245) / false if [none](https://github.com/oriedita/oriedita/commit/6b74945137fa428e8382dfd87946f74f152eeba3#diff-73ccc3efe4a5a8ceb1308fdbcf35773eb47fcc813167228df6d0d0cc4d56cd6fR248) - and I believe this will end up being faster than O(n) / linear time in the best case scenario; albeit the function becomes less reusable as a result

**Comment by TabLand** [comment]
Please note, this change isn't ready for merging yet. 
Undo and Redo buttons should also be in scope for recalculating the selection size...

**Comment by TabLand** [comment]
Hi @Minh-Than,

Could I ask you to give these changes a quick test and gloss over? Please let me know if there's anything stands out in the patch.

I'll try and run a test against Ryujin 3.5 later tonight

**Comment by Minh-Than** [comment]
I tried drawing an orange non interfering aux line and the undo doesn't seem to be enabled. But I tried the yellow version and it seems to work. I think the orange one isn't accounted for, though I'm not exactly sure about getting a solution for this since there's only one set of undo/redo. (I'm not even sure if that's part of the changes you made to be honest haha)

image
image

**Comment by TabLand** [comment]
Hi @Minh-Than  - good catch!

I definitely didn't test with non-interfering aux lines and I hope it's an existing issue. 
I will try to reproduce this issue in this PR as well as against master.

**Comment by TabLand** [comment]
>>I'll try and run a test against Ryujin 3.5 later tonight

Looks like this isn't publicly available in .ori / .fold due to copyright reasons.
Could you advise on the number of creases in the file you tested, specifically "L=[0-9]+" number value that we see on the top left on the canvas?

My system begins to slow down the toggling of selection button state to around 0.5-1 second when L≈17000 (random square grid of mountains and edges), however the UI is also jittery at this point. Things are nearly smooth at L≈15000.

**Comment by TabLand** [comment]
> I tried drawing an orange non interfering aux line and the undo doesn't seem to be enabled. But I tried the yellow version and it seems to work. 

Strangely, I've not been able to replicate this issue on master, issue#325 or the oriedita v1.0.0 installed via debian package; whether orange lines are drawn first or yellow.

Are you able to replicate this issue consistently after cleaning your build?

**Comment by Minh-Than** [comment]
In the ryujin I tested, it’s L=34011.

I’ll try replicating the non aux again, though I’m not exactly sure if the build i downloaded, provided whenever you push, wasn’t updated since I made sure to redownload it before testing out.

**Comment by TabLand** [comment]
> In the ryujin I tested, it’s L=34011.
My system is incredibly jittery at L=32020, however the button state update was still thankfully below 1 second.

**Comment by Minh-Than** [comment]
Ok sorry I've been really busy but I just tried replicating it again using a fresh download a few times and I couldn't replicate them. Seems like a one time thing the first time I tried it. 

And I haven't really find anything else problematic. The issue with keeping up with the performance for L=30k+ seems pretty unlikely as we don't really load up such a huge cp that much, either from those who has really huge density of lines or for testing purposes. Overall it should be fine for now, if there's any suggestion on optimization for any of these then those would be welcomed.

**Comment by TabLand** [comment]
Thanks @Minh-Than - I'm currently happy with this PR as it stands.

@undertrox - appreciate you might be busy; could you review and test when you get a chance or tag another maintainer with "merge into master" access?

**Comment by undertrox** [comment]
Sure, I'll probably get to it later this week. 

By the way, if you want, you can join the discord server that's listed on oriedita.github.io - it's usually a bit quicker to get help/feedback than here. (Also @maks2001916 since they asked about how to get started)

File: oriedita/src/main/java/oriedita/editor/canvas/impl/CreasePattern_Worker_Impl.java
Comment (undertrox) [inline oriedita/src/main/java/oriedita/editor/canvas/impl/CreasePattern_Worker_Impl.java:797]: This line needs to be moved into the if statement to prevent a bug. 
`ApplicationModel.selectPersistent` is a setting that allows you to keep the selection when changing to tools that would usually clear it (e.g. drawing tools, set the Persist checkbox in the bottom left to test it out). The tools still call `unselect_all`, but the actual unselecting is only done if the setting is set to false (or if the caller specifically says to ignore the setting).
 
Currently, if you enable the Persist Checkbox, select something and then select another tool, the buttons get disabled even though the selection isn't empty.

**Review by undertrox — CHANGES_REQUESTED** [review]
looks very good overall, there's just one tiny bug that needs fixing before the merge.

File: oriedita/src/main/java/oriedita/editor/canvas/impl/CreasePattern_Worker_Impl.java
Comment (TabLand) [reply oriedita/src/main/java/oriedita/editor/canvas/impl/CreasePattern_Worker_Impl.java:797]: Aha, good catch, thank you!

I wasn't aware of this button and feature - the existing code makes a lot more sense now.

I'm currently away from my PC for another few days due to Eid holidays. I will patch, test and update this PR as soon as I get back.

**Comment by TabLand** [comment]
Hi @undertrox,
Requested change applied

**Review by undertrox — APPROVED** [review]

**Pull request merged**
